### PR TITLE
TYP: Fix typing in ExtensionDtype registry

### PR DESCRIPTION
**Pull request by Dr-Irv**
Simplifying #40421 to just fix up typing in the registry for Extension Dtypes, and return type for `construct_from_string` in `pandas/core/dtypes/base.py`

### CODE REVIEW

File: pandas/core/dtypes/base.py [hunk 1]
```diff
@@ -213,7 +213,7 @@ def construct_array_type(cls) -> type_t[ExtensionArray]:
         raise NotImplementedError
 
     @classmethod
-    def construct_from_string(cls, string: str):
+    def construct_from_string(cls, string: str) -> ExtensionDtype:
```
Comment (jbrockmendel) [inline pandas/core/dtypes/base.py:216]: can/should we do the `cls: Type[ExtensionDtypeT]` thing here?

File: pandas/core/dtypes/base.py [hunk 2]
```diff
@@ -424,20 +424,23 @@ def register(self, dtype: type[ExtensionDtype]) -> None:
...
 
-    def find(self, dtype: type[ExtensionDtype] | str) -> type[ExtensionDtype] | None:
+    def find(
+        self, dtype: type[ExtensionDtype] | ExtensionDtype | str
+    ) -> type[ExtensionDtype] | ExtensionDtype | None:
         """
         Parameters
         ----------
-        dtype : Type[ExtensionDtype] or str
+        dtype : Type[ExtensionDtype] or ExtensionDtype or str
```
Comment (jbrockmendel) [inline pandas/core/dtypes/base.py:433]: `Type[ExtensionDtype] or ExtensionDtype` -> `ExtensionDtype class or instance`?

File: pandas/core/dtypes/base.py (hunk 1 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: Done there (and in other class methods) in latest commit.

File: pandas/core/dtypes/base.py (hunk 2 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: Done in latest commit.  Also added overloads to clarify behavior as described here https://github.com/pandas-dev/pandas/pull/40421/files#r622387729 which allows a type-ignore to be removed in `pandas/core/dtypes/common.py`

**Review by jreback — APPROVED** [review]

**Comment by Dr-Irv** [comment]
@simonjayhawkins can you review, please?

**Comment by jreback** [comment]
cc @simonjayhawkins if you can (a few open by @Dr-Irv )

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @Dr-Irv. generally lgtm.

File: pandas/core/dtypes/base.py [hunk 3]
```diff
@@ -202,7 +203,7 @@ def names(self) -> list[str] | None:
         return None
 
     @classmethod
-    def construct_array_type(cls) -> type_t[ExtensionArray]:
+    def construct_array_type(cls: type_t[ExtensionDtypeT]) -> type_t[ExtensionArray]:
```
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:206]: done in next commit

File: pandas/core/dtypes/base.py [hunk 4]
```diff
@@ -213,7 +214,9 @@ def construct_array_type(cls) -> type_t[ExtensionArray]:
         raise NotImplementedError
 
     @classmethod
-    def construct_from_string(cls, string: str):
+    def construct_from_string(
+        cls: type_t[ExtensionDtypeT], string: str
+    ) -> ExtensionDtype:
```
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: fixed in next commit

File: pandas/core/dtypes/base.py [hunk 5]
```diff
@@ -424,20 +427,35 @@ def register(self, dtype: type[ExtensionDtype]) -> None:
 
         self.dtypes.append(dtype)
 
-    def find(self, dtype: type[ExtensionDtype] | str) -> type[ExtensionDtype] | None:
+    @overload
+    def find(self, dtype: type[ExtensionDtype]) -> ExtensionDtype:
```
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: fixed in next commit

File: pandas/core/dtypes/base.py [hunk 6]
```diff
@@ -424,20 +427,35 @@ def register(self, dtype: type[ExtensionDtype]) -> None:
...
+    @overload
+    def find(self, dtype: ExtensionDtype) -> ExtensionDtype:
+        ...
+
+    @overload
+    def find(self, dtype: str) -> ExtensionDtype | None:
+        ...
+
+    def find(
+        self, dtype: type[ExtensionDtype] | ExtensionDtype | str
```
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: I checked that and you are right.  I've modified it to accept any type or numpy dtype, and updated the docstring.

**Comment by simonjayhawkins** [comment]
@Dr-Irv removing the milestones on these typing PRs as our types are not public so the 1.3.0rc date is not a concern

**Comment by Dr-Irv** [comment]
@simonjayhawkins  pushed changes a week ago, so looking forward to a new review.

File: pandas/core/dtypes/base.py [hunk 7]
```diff
@@ -424,28 +429,52 @@ def register(self, dtype: type[ExtensionDtype]) -> None:
...
-        dtype : Type[ExtensionDtype] or str
+        dtype : ExtensionDtype class or instance or str or numpy dtype or python type
 
         Returns
         -------
         return the first matching dtype, otherwise return None
         """
         if not isinstance(dtype, str):
-            dtype_type = dtype
+            dtype_type: type[ExtensionDtype] | type
```
Comment (jbrockmendel) [inline pandas/core/dtypes/base.py:463]: what about NpDtype  added to the signature above?

File: pandas/core/dtypes/base.py [hunk 8]
```diff
@@ -424,28 +429,52 @@ def register(self, dtype: type[ExtensionDtype]) -> None:
...
+        self, dtype: NpDtype | type_t[str | float | int | complex | bool | object] | str
+    ) -> ExtensionDtype | None:
+        ...
+
+    def find(
+        self,
+        dtype: type[ExtensionDtype]
+        | ExtensionDtype
+        | NpDtype
+        | type_t[str | float | int | complex | bool | object],
```
Comment (jbrockmendel) [inline pandas/core/dtypes/base.py:451]: arent these subsumed by NpDtype?

File: pandas/core/dtypes/base.py (hunk 7 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:463]: In this PR, `NpDtype` is `Union[str, np.dtype]` .  If `dtype` is `str`, this code isn't entered.  If `dtype` is an `np.dtype`, then an `np.dtype` is a `type`, so adding the `np.dtype` type to the type of `dtype_type` is redundant.

File: pandas/core/dtypes/base.py (hunk 8 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:451]: No, `NpDtype` includes `str`

**Comment by Dr-Irv** [comment]
Merged to master to remove stale label and to see that it passes tests.

**Review by jreback — APPROVED** [review]
lgtm @simonjayhawkins

File: pandas/core/dtypes/base.py [hunk 9]
```diff
@@ -422,28 +427,52 @@ def register(self, dtype: type[ExtensionDtype]) -> None:
 
         self.dtypes.append(dtype)
 
-    def find(self, dtype: type[ExtensionDtype] | str) -> type[ExtensionDtype] | None:
+    @overload
+    def find(self, dtype: type[ExtensionDtypeT]) -> ExtensionDtypeT:
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/base.py:431]:
```
>>> typ
<class 'pandas.core.arrays.integer.Int64Dtype'>
>>> pd.core.dtypes.base.Registry().find(typ)
<class 'pandas.core.arrays.integer.Int64Dtype'>
>>> 
```
```suggestion
    def find(self, dtype: type[ExtensionDtypeT]) -> type[ExtensionDtypeT]:
```

File: pandas/core/dtypes/base.py [hunk 10]
```diff
@@ -422,28 +427,52 @@ def register(self, dtype: type[ExtensionDtype]) -> None:
...
+        ...
+
+    @overload
+    def find(self, dtype: ExtensionDtypeT) -> ExtensionDtypeT:
+        ...
+
+    @overload
+    def find(
+        self, dtype: NpDtype | type_t[str | float | int | complex | bool | object] | str
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/base.py:440]: type_t and type used. choose one for consistency.

File: pandas/core/dtypes/base.py (hunk 8 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/base.py:449]: does #41945 help simplify

File: pandas/core/dtypes/base.py [hunk 11]
```diff
@@ -422,28 +427,52 @@ def register(self, dtype: type[ExtensionDtype]) -> None:
...
+            dtype_type: type[ExtensionDtype] | type
             if not isinstance(dtype, type):
                 dtype_type = type(dtype)
+            else:
+                dtype_type = dtype
             if issubclass(dtype_type, ExtensionDtype):
-                return dtype
+                # cast needed here as mypy doesn't know we have figured
+                # out it is an ExtensionDtype
+                return cast(ExtensionDtype, dtype)
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/base.py:469]: `dtype` could also be `type[ExtensionDtype]` ?

File: pandas/core/dtypes/base.py (hunk 7 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/base.py:461]: the union of `type[ExtensionDtype]` and `type` is just `type`?

File: pandas/core/dtypes/base.py [hunk 12]
```diff
@@ -422,28 +427,52 @@ def register(self, dtype: type[ExtensionDtype]) -> None:
...
 
             return None
 
-        for dtype_type in self.dtypes:
+        for dtype_loop in self.dtypes:
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/base.py:473]: according to the annotations `self.dtypes: list[type[ExtensionDtype]]`

and `dtype_type` is `dtype_type: type[ExtensionDtype] | type`

so why is this changed?

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @Dr-Irv. sorry for the delay in reviewing the latest changes.

File: pandas/core/dtypes/base.py [hunk 13]
```diff
@@ -268,7 +271,7 @@ def construct_from_string(cls, string: str):
         return cls()
 
     @classmethod
-    def is_dtype(cls, dtype: object) -> bool:
+    def is_dtype(cls: type_t[ExtensionDtypeT], dtype: object) -> bool:
```
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: done

File: pandas/core/dtypes/base.py [hunk 14]
```diff
@@ -424,20 +427,35 @@ def register(self, dtype: type[ExtensionDtype]) -> None:
 
         self.dtypes.append(dtype)
 
-    def find(self, dtype: type[ExtensionDtype] | str) -> type[ExtensionDtype] | None:
+    @overload
+    def find(self, dtype: type[ExtensionDtype]) -> ExtensionDtype:
+        ...
+
+    @overload
+    def find(self, dtype: ExtensionDtype) -> ExtensionDtype:
```
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: done

File: pandas/core/dtypes/base.py (hunk 8 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:449]: yes, in next commit

File: pandas/core/dtypes/base.py (hunk 7 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:461]: fixed in next commit

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:431]: Turns out that `type[ExtensionDtypeT]` is not a possible return type, so I removed it.

File: pandas/core/dtypes/base.py (hunk 10 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:440]: changed to `type_t` in next commit

File: pandas/core/dtypes/base.py (hunk 11 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:469]: Not at this point in the logic

File: pandas/core/dtypes/base.py (hunk 12 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:473]: Reverted back due to change of declaration of `dtype_type` above.

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/base.py]: see code sample in comment

File: pandas/core/dtypes/base.py (hunk 11 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/base.py:459]: using code sample from https://github.com/pandas-dev/pandas/pull/41203#discussion_r663202111

```
>>> typ
<class 'pandas.core.arrays.integer.Int64Dtype'>
>>> pd.core.dtypes.base.Registry().find(typ)
> /home/simon/pandas/pandas/core/dtypes/base.py(460)find()
-> return cast(ExtensionDtype, dtype)
(Pdb) dtype
<class 'pandas.core.arrays.integer.Int64Dtype'>
(Pdb) type(dtype)
<class 'type'>
(Pdb) 
```

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: Right, so I can put it back, but then `type[ExtensionDtypeT]` conflicts with `npt.DTypeLike`, because that can be any type, so have to go back to specifying the specific types.

File: pandas/core/dtypes/base.py (hunk 8 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: Had to undo this

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/base.py]: type[object] can create issues. but unfortunately `object` is a valid dtype, use ignores where needed.

File: pandas/core/dtypes/base.py (hunk 11 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:469]: OK, I see.  Fixed in a new commit.

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: > type[object] can create issues. but unfortunately `object` is a valid dtype, use ignores where needed.

So the issue here is that if I use `npt.DTypeLike`, then the signatures with `type_t[ExtensionDtypeT]` and `npt.DTypeLike` overlap, because `npt.DTypeLike` includes `type` in the `Union`.  Are you suggesting to use `npt.DtypeLike` instead of  `NpDtype | type_t[str | float | int | complex | bool | object]` and use `ignore` on the overload overlapping?  Doesn't seem right to me.

File: pandas/core/dtypes/base.py [hunk 15]
```diff
@@ -465,8 +466,8 @@ def find(
                 dtype_type = dtype
             if issubclass(dtype_type, ExtensionDtype):
                 # cast needed here as mypy doesn't know we have figured
-                # out it is an ExtensionDtype
-                return cast(ExtensionDtype, dtype)
+                # out it is an ExtensionDtype or type_t[ExtensionDtype]
+                return cast(Union[ExtensionDtype, type_t[ExtensionDtype]], dtype)
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/base.py:470]: nit, to avoid adding the Union import, i have a preference for,

```suggestion
                return cast("ExtensionDtype | type_t[ExtensionDtype]", dtype)
```

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/base.py]: The goal here is to avoid false positives for users passing a valid dtype and also avoid using our `NpDtype` alias. Ideally we want to remove use of that alias completely.

Examining this function and ignoring the docstring, any type can be passed and the return would be `None` if not an EA type. i.e. this function never raises a TypeError.

maybe widen the types to any, use `object` and if mypy complains use `Any` or ignore?

Comment (simonjayhawkins) [reply pandas/core/dtypes/base.py]: If working with the public api, you may want to consider reviving #40202. 

if not create a test file and post the results (and the test file used) in a similar maner to #40200

File: pandas/core/dtypes/base.py (hunk 15 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:470]: fixed in next commit

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: `Registry.find()` is not in the public API, so do we need to worry about "false positives for users" ?  This code is just so pandas developers don't mess things up.

Comment (simonjayhawkins) [reply pandas/core/dtypes/base.py]: ah yes, The registry attribute was privatized in #40538 so that `pyright --verifytypes` passes. The changes in this PR are not to fix pyright issues?

**Review by simonjayhawkins — COMMENTED** [review]
using 

```python
import numpy as np
import pandas as pd

from pandas.core.dtypes.base import _registry


reveal_type(_registry.find("int"))
reveal_type(_registry.find(object))
reveal_type(_registry.find(np.dtype))
reveal_type(_registry.find(np.dtype("object")))
reveal_type(_registry.find(pd.Int64Dtype))
reveal_type(_registry.find(pd.Int64Dtype()))
```

```
/home/simon/t.py:7: note: Revealed type is "Union[Type[<nothing>], pandas.core.dtypes.base.ExtensionDtype, None]"
/home/simon/t.py:8: note: Revealed type is "Union[Type[<nothing>], pandas.core.dtypes.base.ExtensionDtype, None]"
/home/simon/t.py:9: note: Revealed type is "Union[Type[<nothing>], pandas.core.dtypes.base.ExtensionDtype, None]"
/home/simon/t.py:10: note: Revealed type is "Union[Type[<nothing>], pandas.core.dtypes.base.ExtensionDtype, None]"
/home/simon/t.py:11: note: Revealed type is "Type[pandas.core.arrays.integer.Int64Dtype*]"
/home/simon/t.py:12: note: Revealed type is "pandas.core.arrays.integer.Int64Dtype*"
```

adding another overload for string type (and maybe using npt.DTypeLike) can get us closer?

```
/home/simon/t.py:7: note: Revealed type is "Union[pandas.core.dtypes.base.ExtensionDtype, None]"
/home/simon/t.py:8: note: Revealed type is "None"
/home/simon/t.py:9: note: Revealed type is "None"
/home/simon/t.py:10: note: Revealed type is "None"
/home/simon/t.py:11: note: Revealed type is "Type[pandas.core.arrays.integer.Int64Dtype*]"
/home/simon/t.py:12: note: Revealed type is "pandas.core.arrays.integer.Int64Dtype*"
```

```diff
diff --git a/pandas/core/dtypes/base.py b/pandas/core/dtypes/base.py
index abac3faa97..508c92c671 100644
--- a/pandas/core/dtypes/base.py
+++ b/pandas/core/dtypes/base.py
@@ -17,7 +17,7 @@ import numpy as np
 from pandas._libs.hashtable import object_hash
 from pandas._typing import (
     DtypeObj,
-    NpDtype,
+    npt,
     type_t,
 )
 from pandas.errors import AbstractMethodError
@@ -437,16 +437,21 @@ class Registry:
 
     @overload
     def find(
-        self, dtype: NpDtype | type_t[str | float | int | complex | bool | object]
-    ) -> type_t[ExtensionDtypeT] | ExtensionDtype | None:
+        self, dtype: str
+    ) -> ExtensionDtype | None:
+        ...
+
+    @overload
+    def find(
+        self, dtype: npt.DTypeLike
+    ) -> None:
         ...
 
     def find(
         self,
         dtype: type_t[ExtensionDtype]
         | ExtensionDtype
-        | NpDtype
-        | type_t[str | float | int | complex | bool | object],
+        | npt.DTypeLike
     ) -> type_t[ExtensionDtype] | ExtensionDtype | None:
         """
         Parameters

```

we may in the future decide to create an alias for the literal strings that represent built-in EA dtypes, but not needed here.

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py]: TBH, I don't remember!  It's been since January since I started this adventure.  I did this PR as part of getting `ExtensionArray` working.  I haven't been using `pyright` to verify this.

**Comment by Dr-Irv** [comment]
> adding another overload for string type (and maybe using npt.DTypeLike) can get us closer?

So if you do:
```python
    @overload
    def find(self, dtype: type_t[ExtensionDtypeT]) -> type_t[ExtensionDtypeT]:
        ...

    @overload
    def find(self, dtype: ExtensionDtypeT) -> ExtensionDtypeT:
        ...

    @overload
    def find(self, dtype: str) -> ExtensionDtype | None:
        ...

    @overload
    def find(
        self, dtype: npt.DTypeLike
    ) ->None:
        ...

```
you get this from mypy:
```
pandas\core\dtypes\base.py:431: error: Overloaded function signatures 1 and 4 overlap with incompatible return types  [misc]
pandas\core\dtypes\base.py:439: error: Overloaded function signatures 3 and 4 overlap with incompatible return types  [misc]
```
That's because `npt.DTypeLike` includes `str` and `type` as possible values.  If we get a `str`, we can return `ExtensionDtype` or `None`.  

In the next commit, I changed the 4th overload to be:
```python
    @overload
    def find(
        self, dtype: npt.DTypeLike
    ) -> type_t[ExtensionDtype] | ExtensionDtype | None:
        ...
```

That's because the `npt.DTypeLike` is too wide, and there is not a way to just say "the numpy types" plus "the base types"

**Comment by simonjayhawkins** [comment]
Thanks @Dr-Irv can you post the output of the code sample in https://github.com/pandas-dev/pandas/pull/41203#pullrequestreview-698655949 with the latest changes.

**Comment by Dr-Irv** [comment]
> Thanks @Dr-Irv can you post the output of the code sample in [#41203 (review)](https://github.com/pandas-dev/pandas/pull/41203#pullrequestreview-698655949) with the latest changes.
```
mypytest.py:7: note: Revealed type is "Union[pandas.core.dtypes.base.ExtensionDtype, None]"
mypytest.py:8: note: Revealed type is "Union[Type[pandas.core.dtypes.base.ExtensionDtype], pandas.core.dtypes.base.ExtensionDtype, None]"
mypytest.py:9: note: Revealed type is "Union[Type[pandas.core.dtypes.base.ExtensionDtype], pandas.core.dtypes.base.ExtensionDtype, None]"
mypytest.py:10: note: Revealed type is "Union[Type[pandas.core.dtypes.base.ExtensionDtype], pandas.core.dtypes.base.ExtensionDtype, None]"
mypytest.py:11: note: Revealed type is "Type[pandas.core.arrays.integer.Int64Dtype*]"
mypytest.py:12: note: Revealed type is "pandas.core.arrays.integer.Int64Dtype*"
```

**Comment by Dr-Irv** [comment]
@simonjayhawkins ping...

**Comment by Dr-Irv** [comment]
pinging @simonjayhawkins

**Comment by jreback** [comment]
@Dr-Irv if you can rebase

cc @simonjayhawkins

**Review by simonjayhawkins — APPROVED** [review]
Thanks @Dr-Irv
